client/zeth/zksnark.py

```diff
@@ -78,41 +101,73 @@ def get_contract_name() -> str:
         }
 
     @staticmethod
-    def parse_verification_key(
-            vk_obj: VerificationKey) -> GenericVerificationKey:
+    def verification_key_from_proto(
+            vk_obj: snark_messages_pb2.VerificationKey) -> GenericVerificationKey:
         vk = vk_obj.groth16_verification_key
         return {
-            "alpha_g1": _parse_hex_point_base_group1_affine(vk.alpha_g1),
-            "beta_g2": _parse_hex_point_base_group2_affine(vk.beta_g2),
-            "delta_g2": _parse_hex_point_base_group2_affine(vk.delta_g2),
-            "abc_g1": json.loads(vk.abc_g1),
+            "alpha": group_point_g1_from_proto(vk.alpha_g1),
+            "beta": group_point_g2_from_proto(vk.beta_g2),
+            "delta": group_point_g2_from_proto(vk.delta_g2),
+            "ABC": json.loads(vk.abc_g1),
         }
 
     @staticmethod
-    def parse_proof(proof_obj: ExtendedProof) -> GenericProof:
+    def verification_key_to_proto(
+            vk: GenericVerificationKey) -> snark_messages_pb2.VerificationKey:
+        vk_obj = snark_messages_pb2.VerificationKey()
+        groth16_key = vk_obj.groth16_verification_key  # pylint: disable=no-member
+        group_point_g1_to_proto(vk["alpha"], groth16_key.alpha_g1)
+        group_point_g2_to_proto(vk["beta"], groth16_key.beta_g2)
+        group_point_g2_to_proto(vk["delta"], groth16_key.delta_g2)
+        groth16_key.abc_g1 = json.dumps(vk["ABC"])
+        return vk_obj
+
+    @staticmethod
+    def proof_from_proto(
+            proof_obj: snark_messages_pb2.ExtendedProof) -> GenericProof:
         proof = proof_obj.groth16_extended_proof
         return {
-            "a": _parse_hex_point_base_group1_affine(proof.a),
-            "b": _parse_hex_point_base_group2_affine(proof.b),
-            "c": _parse_hex_point_base_group1_affine(proof.c),
+            "proof": {
+                "a": group_point_g1_from_proto(proof.a),
+                "b": group_point_g2_from_proto(proof.b),
+                "c": group_point_g1_from_proto(proof.c),
+            },
             "inputs": json.loads(proof.inputs),
         }
 
     @staticmethod
-    def mixer_proof_parameters(parsed_proof: GenericProof) -> List[List[Any]]:
+    def proof_to_proto(
+            extproof: GenericProof) -> snark_messages_pb2.ExtendedProof:
+        proof = extproof["proof"]
+        extproof_proto = snark_messages_pb2.ExtendedProof()
+        proof_proto = extproof_proto.groth16_extended_proof \
+            # pylint: disable=no-member
+        group_point_g1_to_proto(proof["a"], proof_proto.a)
+        group_point_g2_to_proto(proof["b"], proof_proto.b)
+        group_point_g1_to_proto(proof["c"], proof_proto.c)
+        proof_proto.inputs = json.dumps(extproof["inputs"])
+        return extproof_proto
+
+    @staticmethod
+    def mixer_proof_parameters(extproof: GenericProof) -> List[List[Any]]:
+        # Note that here we assume a specific form of G2 here, i.e.
+        #   [[X.c0, X.c1], [Y.c0, Y.c1]]
```

Comment:
There is no assumption on the size of the list, or whether the points are in affine form or not.  The assumption is that the coordinates are elements of a proper extension field (i.e. that "b" is a list of lists). I've changed the comment to clarify. It may not be an edge case that we hit in zeth, so may not be a problem.

Reply:
Sorry - I just realised that there is an error in how the assert below is done (basically because the contract interface requires a tuple, but json produces a list, so this assert needs to handle tuples as well). Eventually I'll replace these `GenericProof`, `GenericVerificationKey` types with proper classes and well-defined types.  For now, I'll fix the assert.  Marking as WIP for now.

Reply:
Should be fixed now.

Reply:
> There is no assumption on the size of the list, or whether the points are in affine form or not.

The line `hex_to_int(proof["b"][0] + proof["b"][1]),` is kind of assuming that `b` (in G2) has 2 coordinates (i.e. X, the one accessed by `proof["b"][0]` and Y -> `proof["b"][1]`), no?

Reply:
> Sorry - I just realised that there is an error [...]. Should be fixed now.

Thanks. I also saw that you reworded the comment. Thanks for that

Reply:
> The line hex_to_int(proof["b"][0] + proof["b"][1]), is kind of assuming that b (in G2) has 2 coordinates (i.e. X, the one accessed by proof["b"][0] and Y -> proof["b"][1]), no?

Yes, we definitely assume 2 cordinates.  X and Y. The point here is that we assume that each coordinate is a *list of coefficients*, not a single coefficient.

> As to the extension degree (i.e. size of the list), I interpreted the comment:
[[X.c0, X.c1], [Y.c0, Y.c1]] as an assumption of the structure of the inputs (i.e. each coordinate needs to have 2 coefficients - i.e. we need to manipulate affine points of E'/Fq^2). By looking at the following code this assumption on the degree of the extension is obviously not required.

IIRC, assumption that the extension degree is >1 is implicit in the `+` in `proof["b"][0] + proof["b"][1]` (which assumes it is concatenating lists - and breaks if it concatenates strings).  So your interpretation of the original comment was pretty much right - we assume each coordinate has 2 (or more) coefficients.